// @redwoodjs/cli/dist/lib/configureStorybook.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js/object/define-property");

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault").default;

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.default = extendStorybookConfiguration;

var _path = _interopRequireDefault(require("path"));

var _util = _interopRequireDefault(require("util"));

var _fsExtra = _interopRequireDefault(require("fs-extra"));

var _prettier = _interopRequireDefault(require("prettier"));

var _merge = require("./merge");

var _strategy = require("./merge/strategy");

var _ = require(".");

async function extendStorybookConfiguration(newConfigPath = undefined) {
  const sbPreviewConfigPath = (0, _.getPaths)().web.storybookPreviewConfig;

  if (!_fsExtra.default.existsSync(sbPreviewConfigPath)) {
    await _util.default.promisify(_fsExtra.default.cp)(_path.default.join(__dirname, 'templates', 'storybook.preview.js.template'), sbPreviewConfigPath);
  }

  if (newConfigPath) {
    const read = path => _fsExtra.default.readFileSync(path, {
      encoding: 'utf-8'
    });

    const write = (path, data) => _fsExtra.default.writeFileSync(path, data);

    const merged = (0, _merge.merge)(read(sbPreviewConfigPath), read(newConfigPath), {
      ImportDeclaration: _strategy.interleave,
      ArrayExpression: _strategy.concatUnique,
      ObjectExpression: _strategy.concatUnique,
      ArrowFunctionExpression: _strategy.keepBothStatementParents,
      FunctionDeclaration: _strategy.keepBoth
    });

    const formatted = _prettier.default.format(merged, {
      parser: 'babel',
      ...(await _prettier.default.resolveConfig(sbPreviewConfigPath))
    });

    write(sbPreviewConfigPath, formatted);
  }
}
