// @redwoodjs/api/dist/transforms.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js/object/define-property");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.normalizeRequest = normalizeRequest;
exports.removeNulls = exports.parseEventBody = void 0;

var _crossUndiciFetch = require("cross-undici-fetch");

/**
 * Extracts and parses body payload from event with base64 encoding check
 */
const parseEventBody = event => {
  if (!event.body) {
    return;
  }

  if (event.isBase64Encoded) {
    return JSON.parse(Buffer.from(event.body, 'base64').toString('utf-8'));
  } else {
    return JSON.parse(event.body);
  }
};

exports.parseEventBody = parseEventBody;

function normalizeRequest(event) {
  const body = parseEventBody(event);
  return {
    headers: new _crossUndiciFetch.Headers(event.headers),
    method: event.httpMethod,
    query: event.queryStringParameters,
    body
  };
} // Internal note:  Equivalent to dnull package on npm, which seems to have import issues in latest versions

/**
 * Useful for removing nulls from an object, such as an input from a GraphQL mutation used directly in a Prisma query
 * @param input - Object to remove nulls from
 * See {@link https://www.prisma.io/docs/concepts/components/prisma-client/null-and-undefined Prisma docs: null vs undefined}
 */


const removeNulls = input => {
  for (const key in input) {
    if (input[key] === null) {
      input[key] = undefined;
    } else if (typeof input[key] === 'object' && !(input[key] instanceof Date) // dates are objects too
    ) {
      // Note arrays are also typeof object!
      input[key] = removeNulls(input[key]);
    }
  }

  return input;
};

exports.removeNulls = removeNulls;
